Streamline input output constraints (#137)
ome/ngff-spec: index.md

This means they represent functions from *points* in the input space to *points* in the output space
(see [example below](#spec:example:coordinate_transformation_scale)).

A transform is a JSON object with the following fields:

| Field | Type | Required | Description |
|-------|------|----------|-------------|

> | `type` | string | yes | The type of the transformation, which determines the required and optional fields for that transformation. |

Reviewer: Suggestion to replace the "required" column with a "requirement level" (or synonym), with MUST/SHOULD/MAY to match spec text.

Author: Agree....but that's for a different PR. But, since the requirement level is often context-dependent, I'd rather leave it as a yes/no and then further narrow it in the respective context.

| `name` | string | no | A unique name for this transformation. |

> | `input` | object | yes, unless part of a wrapper transform ([`sequence`](#sequence-md), [`bijection`](#bijection-md), [`byDimension`](#bydimension-md)) | The input coordinate system for this transformation. |

Reviewer: The Type column, where the type is another object defined in this spec, should use that object's name and link to its definition. But we shouldn't try to change all the definitions into tables in this PR :)

Author: It was tempting 😬 🙈

Author: e876295e17b4ecb4dcb8e88e293d9139f77d2ec5

This means they represent functions from *points* in the input space to *points* in the output space
(see [example below](#spec:example:coordinate_transformation_scale)).

A transform is a JSON object with the following fields:

| Field | Type | Required | Description |
|-------|------|----------|-------------|

> | [`type`](transform-type-md)= | string | yes | The type of the transformation, which determines the required and optional fields for that transformation. Possible values |

Reviewer 2: ```suggestion
| [`type`](#transform-type-md) | string | yes | The type of the transformation, which determines the required and optional fields for that transformation. Possible values |
```

:)

(Also, what is "Possible values" about)?

Author: Eh... I think I started writing something and then recondiered. Thanks for the catch 🙈

Author: 11084a30c5f205e9a9a8f437e13e5a05afeab406

| [`bijection`](#bijection-md) | `"forward":Transformation`<br>`"inverse":Transformation` | An invertible transformation providing an explicit forward transformation and its inverse. |
| [`byDimension`](#bydimension-md) | `"transformations":List[Transformation]`.<br>Transformations in the array MUST have<br>`"input_axes": List[number]`, <br> and `"output_axes": List[number]` | A high dimensional transformation using lower dimensional transformations on subsets of dimensions. |

> The parameter values (i.e., `scale` for a [scale transformatiion](#scale-md)) MUST be compatible with input and output space dimensionality (see details). 

Reviewer 2: ```suggestion
The parameter values (e.g., `scale` for a [scale transformatiion](#scale-md)) MUST be compatible with input and output space dimensionality (see details). 
```

  - in the `input` object provide `path` and omit `name`.
  - in the `output` object provide `name` and omit `path`.
  For more information, see [multiscales section below](#multiscales-md).

- **Inside `multiscales > coordinateTransformations`**: Additional transformations for single multiscale images MAY be stored here.
  - One of `input` or `output` MUST reference the "intrinsic" coordinate system by `name` (`path` MAY be omitted or null).

>   - The other MUST reference a named coordinate system in the same multiscales group (by `name`, `path` MAY be omitted or null), or in a child [labels](#labels-md) group (by `name` and `path`).

Reviewer 2: ```suggestion
  - The other MUST reference a named coordinate system either within the same multiscale's `coordinateSystems` (by `name`, `path` MAY be omitted or null), or in a child [labels](#labels-md) group (by `name` and `path`).
```

Easy to forget, but a multiscales group can still contain multiple multiscales...

>   - When referencing a child labels group, the transformation MUST be [`identity`](#identity-md), [`scale`](#scale-md), or [`translation`](#translation-md).

Reviewer 2: Need to check again but aren't labels required to be the same shape as the multiscale; which means this basically always has to be identity? Allowing scale and translation isn't a problem of course, just presents more opportunities for confusion I suppose.

Author: That's the part where we are entering the grey zone between collections and transforms. The relationship between labels and images is first and foremost semantic in nature, not spatial. Hence 0.6 will not force you to add spatial information to your label/image connection, but if you want to, you *can*. 

Hence, the shaping of the labels has never been written out in previous versions and 0.6 won't force this because only collections will express this relationship properly.

Reviewer 2: > The relationship between labels and images is first and foremost semantic in nature, not spatial

Subtle disagreement from the spec itself there:

> Some applications–notably image segmentation–produce a new image that is **in the same coordinate system as a corresponding multiscale image** (usually having the same dimensions and coordinate transformations).

Sounds very much spatial to me :)

Just to be clear, I'm not actually asking for anything to be changed here. Just indulging.

Author: Oh, I see what you mean. The loophole here is "usually"; There was a PR where thought of changing that (possibly #117) and require this transformation to be explicit. Ultimately we decided against it:

- If that transformation were to live in the parent multiscale image, then the addition of a child `labels` image would effectively invalidate the parent image because that transform would then be missing
- If the transform were to live in `/labels/segmentation_image/zarr.json`, its output would have to be `../../zarr.json`, which would *also* be problematic because moving the label image elsewhere would make it invalid and thus no longer self-describing.

The clean option would be to put the label image into a transformation graph and not use the `/labels/...` structure at all. *Or* wait for collections that will make this kind of linkage much more explicit from the point of view of an apex zarr.json in the hierarchy, which would be my preferred solution :)

Reviewer 2: I also don't see any point to requiring the metadata to enumerate all contents. At most I'd go to "SHOULD" level - just like the `labels` metadata only "SHOULD" reference all label multiscales available.

- **Inside `scene > coordinateTransformations`**: Transformations between two or more images.

>   - Both `input` and `output` MUST contain `name`.

Reviewer 2: ```suggestion

- **Inside `scene > coordinateTransformations`**: Transformations between two or more multiscales.
  - Both `input` and `output` MUST specify a coordinate system `name`.
```

: If applications require additional transformations,
  each `multiscales` object MAY contain the field `coordinateTransformations`,
  describing transformations that are applied to all resolution levels in the same manner.
  The following constraints apply:
  - One of `input` or `output` MUST reference the "intrinsic" coordinate system by `name` (`path` MAY be omitted or null).
  - The other of `input` or `output` MUST reference either:

>     - A named coordinate system in the same multiscales group (by `name`, `path` MAY be omitted or null), or

Reviewer 2: ```suggestion
    - A named coordinate system in the same multiscale (by `name`, `path` MAY be omitted or null), or
```
